In [21]:
ls stanfordSentimentTreebank


datasetSentences.txt*  original_rt_snippets.txt*  SOStr.txt*
datasetSplit.txt*      README.txt*                STree.txt*
dictionary.txt*        sentiment_labels.txt*


In [2]:
pathToData = 'stanfordSentimentTreebank/'

In [54]:
id_to_split = {}

with open(pathToData + 'datasetSplit.txt') as F:
    for line in F.readlines()[1:]:
        tmp = line.replace('\n', '').split(',')
        id_to_split[tmp[0]] = tmp[1]
        
id_to_split

{'1': '1',
 '2': '1',
 '3': '2',
 '4': '2',
 '5': '2',
 '6': '2',
 '7': '2',
 '8': '2',
 '9': '2',
 '10': '2',
 '11': '2',
 '12': '2',
 '13': '2',
 '14': '2',
 '15': '2',
 '16': '2',
 '17': '2',
 '18': '2',
 '19': '2',
 '20': '2',
 '21': '2',
 '22': '2',
 '23': '2',
 '24': '2',
 '25': '2',
 '26': '2',
 '27': '2',
 '28': '2',
 '29': '2',
 '30': '2',
 '31': '2',
 '32': '2',
 '33': '2',
 '34': '2',
 '35': '2',
 '36': '2',
 '37': '2',
 '38': '2',
 '39': '2',
 '40': '2',
 '41': '2',
 '42': '2',
 '43': '2',
 '44': '2',
 '45': '2',
 '46': '2',
 '47': '2',
 '48': '2',
 '49': '2',
 '50': '2',
 '51': '2',
 '52': '2',
 '53': '2',
 '54': '2',
 '55': '2',
 '56': '2',
 '57': '2',
 '58': '2',
 '59': '2',
 '60': '2',
 '61': '1',
 '62': '1',
 '63': '1',
 '64': '1',
 '65': '2',
 '66': '2',
 '67': '2',
 '68': '1',
 '69': '2',
 '70': '2',
 '71': '2',
 '72': '1',
 '73': '2',
 '74': '2',
 '75': '2',
 '76': '2',
 '77': '2',
 '78': '2',
 '79': '2',
 '80': '2',
 '81': '2',
 '82': '1',
 '83': '2',
 '84': '2',
 

In [20]:
id_to_val = {}

with open(pathToData + 'sentiment_labels.txt') as FLabels:
    for line in FLabels.readlines()[1:]:
        tmp = line.replace('\n', '').split('|')
        id_to_val[tmp[0]] = tmp[1]
        
id_to_val        

{'0': '0.5',
 '1': '0.5',
 '2': '0.44444',
 '3': '0.5',
 '4': '0.42708',
 '5': '0.375',
 '6': '0.41667',
 '7': '0.54167',
 '8': '0.33333',
 '9': '0.45833',
 '10': '0.47222',
 '11': '0.59722',
 '12': '0.33333',
 '13': '0.93056',
 '14': '0.80556',
 '15': '0.81944',
 '16': '0.76389',
 '17': '0.5',
 '18': '0.5',
 '19': '0.69444',
 '20': '0.5',
 '21': '0.75',
 '22': '0.29167',
 '23': '0.31944',
 '24': '0.73611',
 '25': '0.77778',
 '26': '0.59722',
 '27': '0.58333',
 '28': '0.68056',
 '29': '0.48611',
 '30': '0.52778',
 '31': '0.23611',
 '32': '0.29167',
 '33': '0.56944',
 '34': '0.375',
 '35': '0.47222',
 '36': '0.29167',
 '37': '0.375',
 '38': '0.30556',
 '39': '0.52778',
 '40': '0.18056',
 '41': '0.61111',
 '42': '0.59722',
 '43': '0.72222',
 '44': '0.33333',
 '45': '0.52778',
 '46': '0.40278',
 '47': '0.22222',
 '48': '0.31944',
 '49': '0.29167',
 '50': '0.66667',
 '51': '0.61111',
 '52': '0.43056',
 '53': '0.77778',
 '54': '0.40278',
 '55': '0.55556',
 '56': '0.55556',
 '57': '0.76389',

In [41]:
dictionary  = {}

with open(pathToData + 'dictionary.txt') as FDict:
    for line in FDict.readlines():
        tmp = line.replace('\n', '').split('|')
        dictionary[tmp[0]] = (tmp[1], id_to_val[tmp[1]])

dictionary

{'!': ('0', '0.5'),
 "! '": ('22935', '0.52778'),
 "! ''": ('18235', '0.5'),
 '! Alas': ('179257', '0.44444'),
 '! Brilliant': ('22936', '0.86111'),
 '! Brilliant !': ('40532', '0.93056'),
 "! Brilliant ! '": ('22937', '1'),
 "! C'mon": ('60624', '0.47222'),
 "! Gollum 's ` performance ' is incredible": ('13402', '0.76389'),
 '! Oh , look at that clever angle ! Wow , a jump cut !': ('179258',
  '0.27778'),
 '! Romething': ('140882', '0.5'),
 '! Run': ('179259', '0.43056'),
 '! The Movie': ('60625', '0.5'),
 '! The camera twirls ! Oh , look at that clever angle ! Wow , a jump cut !': ('179260',
  '0.22222'),
 '! True Hollywood Story': ('140883', '0.55556'),
 '! Wow': ('179261', '0.77778'),
 '! Zoom !': ('179262', '0.63889'),
 '!?': ('220445', '0.5'),
 "!? '": ('220446', '0.5'),
 '#': ('60626', '0.5'),
 '# 133': ('179263', '0.5'),
 '# 3': ('220447', '0.5'),
 '# 8217': ('140884', '0.48611'),
 '# 8217 ;': ('140885', '0.5'),
 '# 8217 ; t': ('140886', '0.5'),
 '# 9': ('60627', '0.58333'),
 '

In [42]:
data = {}

with open(pathToData + 'datasetSentences.txt') as FDSentenses:
    for line in FDSentenses.readlines()[1:]:
        tmp = line.replace('\n', '').split('\t')
        if tmp[1] in dictionary_v2:
            data[tmp[1]] = (tmp[0], dictionary_v2[tmp[1]][1])
        
data

{"The Rock is destined to be the 21st Century 's new `` Conan '' and that he 's going to make a splash even greater than Arnold Schwarzenegger , Jean-Claud Van Damme or Steven Segal .": ('1',
  '0.69444'),
 "The gorgeously elaborate continuation of `` The Lord of the Rings '' trilogy is so huge that a column of words can not adequately describe co-writer\\/director Peter Jackson 's expanded vision of J.R.R. Tolkien 's Middle-earth .": ('2',
  '0.83333'),
 'Effective but too-tepid biopic': ('3', '0.51389'),
 'If you sometimes like to go to the movies to have fun , Wasabi is a good place to start .': ('4',
  '0.73611'),
 "Emerges as something rare , an issue movie that 's so honest and keenly observed that it does n't feel like one .": ('5',
  '0.86111'),
 'The film provides some great insight into the neurotic mindset of all comics -- even those who have reached the absolute top of the game .': ('6',
  '0.59722'),
 'Offers that rare combination of entertainment and education .': ('7',
 

In [37]:
len(data)

11272

In [46]:
from collections import namedtuple

In [101]:
import gensim
from gensim.models.doc2vec import TaggedDocument
from gensim.models import Doc2Vec
import multiprocessing

In [131]:
class myClass:
    def __init__(self, words, tags, val):
        self.words = words
        self.tags = tags
        self.val = val

In [447]:
train_val_test_precise = [[], [], []]
train_val_test = [[], [], []]

to_class_precise = [0, 1, 2, 3, 4]

security = 1e-4

for sentence in data:
    val = data[sentence]
    tag = val[0]
    split = int(id_to_split[tag]) - 1
    sentiment = float(val[1])
    words = gensim.utils.to_unicode(sentence).split()
    
    # Well defined Classes sets
    myC = myClass(words, [tag],
                  to_class_precise[int(sentiment * (len(to_class) - security))])
    train_val_test_precise[split].append(myC)
    
    # Add in Positive/Negativ sets
    myC = myClass(words, [tag],
                  int(sentiment >= 0.5))
    train_val_test[split].append(myC)

In [448]:
train = train_val_test[0]
val = train_val_test[1]
test = train_val_test[2]

train_precise = train_val_test_precise[0]
val_precise = train_val_test_precise[1]
test_precise = train_val_test_precise[2]

In [449]:
 train[0].tags, train[0].val, train[0].words

(['1'],
 1,
 ['The',
  'Rock',
  'is',
  'destined',
  'to',
  'be',
  'the',
  '21st',
  'Century',
  "'s",
  'new',
  '``',
  'Conan',
  "''",
  'and',
  'that',
  'he',
  "'s",
  'going',
  'to',
  'make',
  'a',
  'splash',
  'even',
  'greater',
  'than',
  'Arnold',
  'Schwarzenegger',
  ',',
  'Jean-Claud',
  'Van',
  'Damme',
  'or',
  'Steven',
  'Segal',
  '.'])

## Model part:

In [394]:
# Usefull to evaluate evolution of our errors

from collections import defaultdict
error_rates = defaultdict(lambda: 1.0)  # To selectively print only best errors achieved

In [403]:
# Build model

cores = multiprocessing.cpu_count()


model = Doc2Vec(dm=1, vector_size=400,
                window=8, negative=5, hs=0, min_count=1, sample=0,
                epochs=20, workers=cores, alpha=0.05,
                comment='basic')

In [404]:
# Build vocabulary

model.build_vocab(train)

In [405]:
# Train Model

model.train(train, total_examples=len(train),
            epochs=model.epochs)

In [406]:
def vec_for_learning(model, tagged_docs):
    targets, regressors = zip(*[(doc.val, model.infer_vector(doc.words)) for doc in tagged_docs])
    return targets, regressors

In [451]:
from sklearn.linear_model import LogisticRegression

def eval_gensim(model):
    y_train, X_train = vec_for_learning(model, train)
    y_val, X_val = vec_for_learning(model, val)

    logreg = LogisticRegression(n_jobs=1, C=1e5)
    logreg.fit(X_train, y_train)
    y_pred = logreg.predict(X_val)
    
    return accuracy_score(y_val, y_pred)


accuracy = eval_gensim(model)

print('Testing accuracy %s' % accuracy)

/home/dylan/.local/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Testing accuracy 0.6980687706076307


In [408]:
error_rates[str(model)] = accuracy

In [409]:
error_rates

defaultdict(<function __main__.<lambda>()>,
            {'Doc2Vec("basic",dbow,d400,n5,t4)': 0.7027790861987753,
             'Doc2Vec("basic",dm/m,d400,n5,w8,t4)': 0.7046632124352331})

### Improvement: ConcatenatedDoc2Vec

In [440]:
from gensim.test.test_doc2vec import ConcatenatedDoc2Vec

In [456]:
subModel1 = Doc2Vec(dm=1, vector_size=400,
                window=8, negative=5, hs=0, min_count=1, sample=0,
                epochs=20, workers=cores, alpha=0.05,
                comment='basic')
subModel2 = Doc2Vec(dm=0, vector_size=400,
                window=8, negative=5, hs=0, min_count=1, sample=0,
                epochs=20, workers=cores, alpha=0.05,
                comment='basic')

In [457]:
subModel1.build_vocab(train)
subModel2.build_vocab(train)

In [458]:
subModel1.train(train, total_examples=len(train),
            epochs=subModel1.epochs)
subModel2.train(train, total_examples=len(train),
            epochs=subModel2.epochs)

In [459]:
# Both models need to be trained on their own before
concatModel = ConcatenatedDoc2Vec([subModel1, subModel2])

In [460]:
accuracy = eval_gensim(concatModel)

print('Testing accuracy %s' % accuracy)

/home/dylan/.local/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Testing accuracy 0.7150259067357513


In [461]:
error_rates[str(concatModel)] = accuracy
error_rates

defaultdict(<function __main__.<lambda>()>,
            {'Doc2Vec("basic",dbow,d400,n5,t4)': 0.7027790861987753,
             'Doc2Vec("basic",dm/m,d400,n5,w8,t4)': 0.7046632124352331,
             "CountVectorizer(analyzer='word', bi": 0.6759302873292511,
             'Doc2Vec("basic",dm/m,d400,n5,w8,t4)+Doc2Vec("basic",dbow,d400,n5,t4)': 0.7150259067357513})

Both models had respectively 
- *dbow* : 0.7027790861987753
- *dm/m* : 0.70466321243523

By combining them we've improved to : **0.7150259067357513**

## Comparing to other methods

### Bag of Words:

In [ ]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer

from sklearn import metrics


dbow_train_val_test_precise = [[], [], []]
dbow_train_val_test = [[], [], []]

to_class_precise = [0, 1, 2, 3, 4]

security = 1e-4

for sentence in data:
    val = data[sentence]
    tag = val[0]
    split = int(id_to_split[tag]) - 1
    sentiment = float(val[1])
    words = sentence
    
    # Well defined Classes sets
    myC = myClass(words, [tag],
                  to_class_precise[int(sentiment * (len(to_class) - security))])
    dbow_train_val_test_precise[split].append(myC)
    
    # Add in Positive/Negativ sets
    myC = myClass(words, [tag],
                  int(sentiment >= 0.5))
    dbow_train_val_test[split].append(myC)
    
    
# All data
newsgroups_train_data = [s.words for s in dbow_train_val_test[0]]
newsgroups_test_data = [s.words for s in dbow_train_val_test[1]]

newsgroups_train_target = [s.val for s in dbow_train_val_test[0]]
newsgroups_test_target = [s.val for s in dbow_train_val_test[1]]

In [436]:

# Create our vectorizer
vectorizer = CountVectorizer()

# Get the training vectors
vectors = vectorizer.fit_transform(newsgroups_train_data)

# Build the classifier
logreg = LogisticRegression(n_jobs=1, C=1e5)

#  Train the classifier
logreg.fit(vectors, newsgroups_train_target)

# Get the test vectors
vectors_test = vectorizer.transform(newsgroups_test_data)

# Predict and score the vectors
pred = logreg.predict(vectors_test)
acc_score = metrics.accuracy_score(newsgroups_test_target, pred)
f1_score = metrics.f1_score(newsgroups_test_target, pred, average='macro')

print('Total accuracy classification score: {}'.format(acc_score))
print('Total F1 classification score: {}'.format(f1_score))

/home/dylan/.local/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Total accuracy classification score: 0.6759302873292511
Total F1 classification score: 0.6742825234906992


In [439]:
error_rates[str(vectorizer)[:35]] = acc_score
error_rates

defaultdict(<function __main__.<lambda>()>,
            {'Doc2Vec("basic",dbow,d400,n5,t4)': 0.7027790861987753,
             'Doc2Vec("basic",dm/m,d400,n5,w8,t4)': 0.7046632124352331,
             "CountVectorizer(analyzer='word', bi": 0.6759302873292511})